openhcl/virt_mshv_vtl/src/processor/hardware_cvm/mod.rs

```diff
@@ -607,6 +608,72 @@ impl<B: HardwareIsolatedBacking> UhProcessor<'_, B> {
 
         Ok(())
     }
+
+    pub(crate) fn hcvm_handle_cross_vtl_interrupts(
+        &mut self,
+        interrupt_pending: VtlArray<Option<u8>, 2>,
+        lapic_msr_read: impl FnOnce(&mut Self, GuestVtl, u32) -> Result<u64, MsrError>,
+    ) {
+        match self.backing.cvm_state_mut().exit_vtl {
+            GuestVtl::Vtl0 => {
+                // Check for VTL preemption
+                if let Some(vector) = interrupt_pending[GuestVtl::Vtl1] {
+                    let priority = vector >> 4;
+                    if priority > 0 {
+                        let ppr = lapic_msr_read(
+                            self,
+                            GuestVtl::Vtl1,
+                            ApicRegister::PPR.0 as u32 + x86defs::apic::X2APIC_MSR_BASE,
+                        )
+                        .expect("reading PPR should never fail");
+                        let ppr_priority = ppr >> 4;
+                        if priority as u64 > ppr_priority {
+                            B::switch_vtl_state(self, GuestVtl::Vtl0, GuestVtl::Vtl1);
+                            self.backing.cvm_state_mut().exit_vtl = GuestVtl::Vtl1;
+                            self.hv[GuestVtl::Vtl1]
+                                .as_ref()
+                                .unwrap()
+                                .set_return_reason(HvVtlEntryReason::INTERRUPT)
+                                .unwrap();
+                        }
+                    }
+                }
+            }
+            GuestVtl::Vtl1 => {
+                // Check for VINA
+                if let Some(vector) = interrupt_pending[GuestVtl::Vtl0] {
+                    let priority = vector >> 4;
+                    if priority > 0 {
+                        let hv = self.hv[GuestVtl::Vtl1].as_mut().unwrap();
+                        if hv.synic.vina().enabled() && !hv.vina_asserted().unwrap() {
+                            let ppr = lapic_msr_read(
+                                self,
+                                GuestVtl::Vtl1,
+                                ApicRegister::PPR.0 as u32 + x86defs::apic::X2APIC_MSR_BASE,
+                            )
+                            .expect("reading PPR should never fail");
+                            let ppr_priority = ppr >> 4;
+                            if priority as u64 > ppr_priority {
+                                let vp_index = self.vp_index();
+                                let hv = self.hv[GuestVtl::Vtl1].as_mut().unwrap();
+                                hv.set_vina_asserted(true).unwrap();
+                                self.partition
```

Comment:
We are supposed to inject the interrupt only if VINA was not previously asserted. See <https://learn.microsoft.com/en-us/virtualization/hyper-v-on-windows/tlfs/vsm#virtual-interrupt-notification-assist> for the documented semantics.

Reply:
Oh, you check that above. OK.